plaso/parsers/postgresql.py

```diff
@@ -0,0 +1,190 @@
+# -*- coding: utf-8 -*-
+"""Parser for PostgreSQL application log files."""
+
+import string
+import pyparsing
+
+from dateutil import parser
+
+from dfdatetime import time_elements as dfdatetime_time_elements
+
+from plaso.containers import events
+from plaso.containers import time_events
+from plaso.lib import definitions
+from plaso.lib import errors
+from plaso.parsers import manager
+from plaso.parsers import text_parser
+
+class PostgreSQLEventData(events.EventData):
+  """PostgreSQL application log data.
+
+  Attributes:
+    log_level (str): logging level of event
+    log_line (str): log message.
+    pid (int): process identifier (PID).
+    user (str): "user@database" string if present.
+        Records the user account and database name
+        that was authenticated or attempting to authenticate.
+  """
+
+  DATA_TYPE = 'postgresql:application_log:entry'
+
+  def __init__(self):
+    """Initializes event data."""
+    super(PostgreSQLEventData, self).__init__(data_type=self.DATA_TYPE)
+    self.log_level = None
+    self.log_line = None
+    self.pid = None
+    self.user = None
+
+
+class PostgreSQLParser(text_parser.PyparsingMultiLineTextParser):
+  """Parses events from PostgreSQL application log files.
+
+  This is a multi-line log format that records internal database
+  application logs as well as authentication attempts.
+
+  """
+
+  NAME = 'postgresql'
+  DATA_FORMAT = 'PostgreSQL application log file'
+  _ENCODING = 'utf-8'
+
+  _DATE_TIME = pyparsing.Group(
+      pyparsing.LineStart() +
+      text_parser.PyparsingConstants.DATE_ELEMENTS +
+      text_parser.PyparsingConstants.TIME.setResultsName('time') +
+      pyparsing.Optional(
+          pyparsing.Suppress('.') +
+          text_parser.PyparsingConstants.INTEGER
+      ).setResultsName('microseconds') +
```

Comment:
per spec is this always milliseconds (3 digits)?

Reply:
Done

Reply:
(Correct, although it's not required to have it at all)